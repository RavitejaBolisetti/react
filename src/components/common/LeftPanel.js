/*
 *   Copyright (c) 2023 Mahindra & Mahindra Ltd.
 *   All rights reserved.
 *   Redistribution and use of any source or binary or in any form, without written approval and permission is prohibited. Please read the Terms of Use, Disclaimer & Privacy Policy on https://www.mahindra.com/
 */
import { Tree } from 'antd';
import { useMemo, useState, useEffect } from 'react';
import styles from './TreeView.module.css';

const LeftPanel = (props) => {
    const { selectedTreeKey, treeData, fieldNames, handleTreeViewClick, isOpenInModal, checkedKeys, expendedKeys: defaultExpandedKeys = [] } = props;
    console.log('ðŸš€ ~ file: LeftPanel.js:12 ~ LeftPanel ~ defaultExpandedKeys:', defaultExpandedKeys);
    const { isTreeViewVisible, checkable, onCheck = () => {} } = props;
    const { searchValue, setSearchValue } = props;
    const { defaultCheckedKeys = [], defaultSelectedKeys = [], disableCheckbox = false, noLeftRightPadding = false } = props;

    const [expandedKeys, setExpandedKeys] = useState([]);
    const [autoExpandParent, setAutoExpandParent] = useState(true);
    console.log('ðŸš€ ~ file: LeftPanel.js:19 ~ LeftPanel ~ autoExpandParent:', autoExpandParent);

    const onExpand = (newExpandedKeys) => {
        setExpandedKeys(newExpandedKeys);
        setAutoExpandParent(false);
    };

    const dataList = [];
    const generateList = (data) => {
        for (let node of data) {
            dataList.push({
                id: node[fieldNames?.key],
                title: node[fieldNames?.title],
            });
            if (node[fieldNames?.children]) {
                generateList(node[fieldNames?.children]);
            }
        }
    };

    treeData && generateList(treeData);
    const getParentKey = (key, tree) => {
        let parentKey;
        for (let i = 0; i < tree.length; i++) {
            const node = tree[i];
            if (node[fieldNames?.children]) {
                if (node[fieldNames?.children].some((item) => item[fieldNames?.key] === key)) {
                    parentKey = node[fieldNames?.key];
                } else if (getParentKey(key, node[fieldNames?.children])) {
                    parentKey = getParentKey(key, node[fieldNames?.children]);
                }
            }
        }
        return parentKey;
    };

    useEffect(() => {
        const newExpandedKeys = dataList
            ?.map((item) => {
                if (item?.title?.toLowerCase()?.indexOf(searchValue?.toLowerCase()) > -1) {
                    return getParentKey(item?.id, treeData);
                }
                return null;
            })
            .filter((item, i, self) => item && self?.indexOf(item) === i);
        setExpandedKeys(searchValue ? newExpandedKeys : []);
        setSearchValue(searchValue);
        setAutoExpandParent(true);
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [searchValue]);

    // const panelParentClass = isTreeViewVisible ? styles.panelVisible : styles.panelHidden;

    const finalTreeData = useMemo(() => {
        const loop = (data) =>
            data?.map((item) => {
                const strTitle = item[fieldNames?.title];

                const strTitleLowerCase = strTitle?.toLowerCase();
                const searchValueLowerCase = searchValue?.toLowerCase();
                const index = strTitleLowerCase?.indexOf(searchValueLowerCase);

                const beforeStr = strTitle?.substring(0, index);
                const afterStr = strTitle?.slice(index + searchValue?.length);
                const title =
                    index > -1 ? (
                        <span>
                            {beforeStr}
                            <span className="site-tree-search-value" style={{ color: 'red' }}>
                                {strTitle?.substring(index, index + searchValue?.length)}
                            </span>
                            {afterStr}
                        </span>
                    ) : (
                        <span>{strTitle}</span>
                    );
                if (item[fieldNames?.children]) {
                    return {
                        title,
                        disableCheckbox,
                        key: item[fieldNames?.key],
                        children: loop(item[fieldNames?.children]),
                    };
                }
                return {
                    title,
                    disableCheckbox,
                    key: item[fieldNames?.key],
                };
            });
        return loop(treeData);
    }, [searchValue, fieldNames, treeData, disableCheckbox]);

    const noLeftRightPaddingParentClass = noLeftRightPadding && styles.noLeftRightParentPadding;
    const noLeftRightPaddingClass = noLeftRightPadding && styles.noLeftRightPadding;

    return (
        <div>
            {isTreeViewVisible ? (
                <div className={isOpenInModal ? styles.modalView : ''}>
                    <div className={styles.scrollTreeData}>
                        <Tree onCheck={onCheck} defaultExpandedKeys={['SAMGC1', 'SAMCO1']} defaultCheckedKeys={['SAMGC1', 'SAMCO1']} defaultSelectedKeys={['SAMGC1', 'SAMCO1']} checkedKeys={checkedKeys} checkable={checkable} onSelect={handleTreeViewClick} showLine={true} showIcon={true} onExpand={onExpand} treeData={finalTreeData} />
                    </div>
                </div>
            ) : undefined}
        </div>
    );
};
export default LeftPanel;
