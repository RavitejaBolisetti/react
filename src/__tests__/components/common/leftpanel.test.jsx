/*
 *   Copyright (c) 2023 Mahindra & Mahindra Ltd.
 *   All rights reserved.
 *   Redistribution and use of any source or binary or in any form, without written approval and permission is prohibited. Please read the Terms of Use, Disclaimer & Privacy Policy on https://www.mahindra.com/
 */
import { render, fireEvent, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import LeftPanel from '@components/common/LeftPanel';
import userEvent from '@testing-library/user-event';
const treeData =[
    {
        shortDescription: 'parent 1',
        id: 'parent 1',
        children: [
            {
                shortDescription: 'asian',
                id: 'AS',

                children: [
                    {
                        shortDescription: 'India',
                        id: 'IND',
                    },
                    {
                        title: 'node2',
                        key: 'node2',

                        children: [
                            {
                                title: 'randomNode_2',
                                key: 'randomNode_2',

                                children: [
                                    {
                                        title: 'node2',
                                        key: 'node2',

                                        children: [
                                            {
                                                title: 'randomNode_3',
                                                key: 'randomNode_3',
                                            },
                                        ],
                                    },
                                ],
                            },
                        ],
                    },
                ],
            },
        ],
    },
    {
        shortDescription: 'topNode',
        id: 'topNode',
        children: [
            {
                title: 'node1',
                key: 'node1',

                children: [
                    {
                        title: 'randomNode_1',
                        key: 'randomNode_1',
                    },
                    {
                        title: 'node2',
                        key: 'node2',

                        children: [
                            {
                                title: 'randomNode_2',
                                key: 'randomNode_2',

                                children: [
                                    {
                                        title: 'node2',
                                        key: 'node2',

                                        children: [
                                            {
                                                title: 'randomNode_3',
                                                key: 'randomNode_3',
                                            },
                                        ],
                                    },
                                ],
                            },
                        ],
                    },
                ],
            },
        ],
    },
];
describe('LeftPanel', () => {
  it('renders panelVisible when isTreeViewVisible is true', () => {
    const container = render(<LeftPanel isTreeViewVisible={true} isOpenInModal={true}  searchValue="Finac" setSearchValue={jest.fn()} setExpandedKeys={jest.fn()} onExpand={jest.fn()} autoExpandParent={true} setAutoExpandParent={true} />);
    const component = document.querySelector('.treeViewContainer');
    expect(component).toHaveClass('treeViewContainer');
  });

  it('shows tree nodes with matching search value in red', () => {
    const {container  } = render(<LeftPanel isTreeViewVisible={true} treeData={treeData} searchValue="Finac" setSearchValue={jest.fn()} isOpenInModal={true} setExpandedKeys={jest.fn()} onExpand={jest.fn()} autoExpandParent={false}  setAutoExpandParent={jest.fn()} />);
    expect(container).toMatchSnapshot();
  });
});