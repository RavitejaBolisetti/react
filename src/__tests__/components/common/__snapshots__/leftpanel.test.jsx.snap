// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeftPanel shows tree nodes with matching search value in red 1`] = `
<div>
  <div>
    <div
      class="modalView"
    >
      <div
        class="scrollTreeData"
      >
        <div
          class="ant-tree css-dev-only-do-not-override-14wwjjs ant-tree-show-line"
          role="tree"
        >
          <div>
            <input
              aria-label="for screen reader"
              style="width: 0px; height: 0px; display: flex; overflow: hidden; opacity: 0; border: 0px; padding: 0px; margin: 0px;"
              tabindex="0"
              value=""
            />
          </div>
          <div
            aria-hidden="true"
            class="ant-tree-treenode"
            style="position: absolute; pointer-events: none; visibility: hidden; height: 0px; overflow: hidden; border: 0px; padding: 0px;"
          >
            <div
              class="ant-tree-indent"
            >
              <div
                class="ant-tree-indent-unit"
              />
            </div>
          </div>
          <div
            class="ant-tree-list"
            style="position: relative;"
          >
            <div
              class="ant-tree-list-holder"
            >
              <div>
                <div
                  class="ant-tree-list-holder-inner"
                  style="display: flex; flex-direction: column;"
                >
                  <div
                    aria-grabbed="false"
                    class="ant-tree-treenode ant-tree-treenode-switcher-close"
                    draggable="false"
                  >
                    <span
                      aria-hidden="true"
                      class="ant-tree-indent"
                    />
                    <span
                      class="ant-tree-switcher ant-tree-switcher-noop"
                    >
                      <span
                        class="ant-tree-switcher-leaf-line"
                      />
                    </span>
                    <span
                      class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-normal"
                      title=""
                    >
                      <span
                        class="ant-tree-iconEle ant-tree-icon__docu"
                      />
                      <span
                        class="ant-tree-title"
                      >
                        <span />
                      </span>
                    </span>
                  </div>
                  <div
                    aria-grabbed="false"
                    class="ant-tree-treenode ant-tree-treenode-switcher-close ant-tree-treenode-leaf-last"
                    draggable="false"
                  >
                    <span
                      aria-hidden="true"
                      class="ant-tree-indent"
                    />
                    <span
                      class="ant-tree-switcher ant-tree-switcher-noop"
                    >
                      <span
                        class="ant-tree-switcher-leaf-line"
                      />
                    </span>
                    <span
                      class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-normal"
                      title=""
                    >
                      <span
                        class="ant-tree-iconEle ant-tree-icon__docu"
                      />
                      <span
                        class="ant-tree-title"
                      >
                        <span />
                      </span>
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
